### PR TITLE
[NOREF] Implement polymorphic dictionary (keyed on strings)

### PR DESCRIPTION
**Pull request by Per48edjes**
Took a first stab at implementing a basic, dynamically-sized, open-addressed dictionary (a subset of the Set ADT) using a (non-cryptographic) hashing function I found online and a linear probing strategy to deal with collisions.

Keeping this as a draft so we can discuss live tomorrow, but definitely feel free to add any comments in the meantime!

### CODE REVIEW

File: lib/collections.c
Comment (Per48edjes) [inline lib/collections.c]: Figured this might be useful for the caller of the dictionary interface functions since we're throwing errors when keys don't exist (example usage: `if (dictionary_contains(...)) { dictionary_remove(...) }`)

File: lib/collections.c
Comment (Per48edjes) [inline lib/collections.c]: Thinking we might need to make a copy of the key string here (and eventually free it in `dictionary_free`)

**Comment by khollbach** [comment]
Nice! Great implementation.

Good call on free-ing keys. We can also leave this as a TODO for once we have code that uses this API; might be easier to reason about / make decisions then.

I've scribbled down some notes which we can look at in-person -- mostly ideas about refactoring the probing code into a helper function.

**Review by khollbach — APPROVED** [review]

**Comment by Per48edjes** [comment]
@khollbach , let me know your thoughts on aa79048 ! Namely, I could use another set of 👀 s on:

- `dictionary_rebuild` (now taking as input a `Dictionary**`) leveraging `dictionary_set`; let me know if I've properly updated the dictionary to be the "new" dictionary created therein
- `dictionary_lookup` is an internal function to do the linear probing
- Confirming that we shoudl be `free`-ing the existing `key` (since we copy the `char*` on insertion) in `dictionary_remove`

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c:309]: `contains` can be refactored in terms of `lookup` as well; ~`return lookup() >= 0;`

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c]: you can move this to the `.h` file and remove the prototype for the old `insert` method

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c]: This is pretty subtle, but I think having a `Dictionary**` in a helper method while the main API takes `Dictionary*` won't work.

---

If a user of the library has a local variable `Dictionary *d` pointing to an existing dict, and they call a method that resizes (say, `remove`), they're calling it as `dict_remove(d)`.

This means that no matter what happens, the local variable in that user code will stay pointing to the same `Dictionary`. But in the implementation of `remove` when you `rebuild`, you're modifying the `Dictionary*` *local variable of rebuild* to point to the new `Dictionary` -- not the caller's local variable.

Hopefully that makes sense, but no worries if it doesn't; this explanation is probably a bit incomprehensible without a diagram ^^'

lmk what you think

---

And as far as what to do instead, I also have some thoughts, so lmk if you'd like to hear them!

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c]: can delete this line

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c]: I think this condition is *implied* by the spec of `dictionary_lookup`, so it should be unnecessary.

optional: you could make it an assert, to signal intent to the reader

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c]: again, string equality should be implied in this case

File: lib/collections.c
Comment (khollbach) [inline lib/collections.c]: same here, strcmp not needed

**Review by khollbach — COMMENTED** [review]
Left some comments; I think the main annoyance will be fixing the issue with rebuild, but I contend that it should be possible. (In particular, you should be able to mutate the key_value_pairs field of the existing dictionary to point to the new allocation.)

Also, if you prefer to keep the original impl of rebuild, that's fine too!

File: lib/collections.c
Comment (Per48edjes) [reply lib/collections.c]: Hmmmmm...let's definitely chat through this on Wednesday or Thursday live. (Will hold off on making any substantive changes until we can sync IRL.)

Comment (khollbach) [reply lib/collections.c]: Sure thing!

I'll do the same. Still planning to add nice docs to things, for ownership contracts and all that -- but I'm hoping to avoid merge conflicts by waiting until this change lands on main.

**Pull request merged**
